### DISCUSSION TITLE
how to deactivate MagicNumber issues for 'when' branches in detekt.yml

### DISCUSSION THREAD

**Discussion by christian-draeger** [detekt/detekt#3796] [Q&A] [answered]
i got MagicNumber issues if i have a when (switch case) expression like this:

```kotlin
when (x) {
    42 -> print("x is 42")
    47, 11 -> print("x is 47 or 11")
    in 1..10 -> print("x is in the range")
    !in 10..20 -> print("x is outside the range")
    else -> print("none of the above")
}
``` 

or here a real world example

![image](https://user-images.githubusercontent.com/22223458/118788390-0b593b80-b894-11eb-9536-d90401eed889.png)

since i think this is a valid usecase (at least in my current project where i have to decide stuff based on train lines^^) i want to disable the MagicNumber rule for switch cases for the whole project via detekt.yml but could not find how

**Comment by marschwar** [comment, marked as answer]
I don't know that I agree with the assessment that this would be a valid use case where those numbers should not be considered magic numbers.  Maybe there is a different way to model your domain?

But to answer your question; your are right, there is no option to disable magic number detection for when expressions. You can of course disable the rule altogether or add `@Suppress("MagicNumber")` to the expression.

**Reply by christian-draeger** [reply]
i think it would be helpful to have at least the possibility to decide. i think it shouldn't be a magic number in the when case but for sure don't want to disable the magic number rule completely.  being forced to annotate all the occurrences with `@Suppress` would feel odd to me as well since i have multiple of such in my project.

**Reply by marschwar** [reply]
But what exactly makes the when expression different so that it should be allowed to use magic numbers here? I realize that it would be convenient in your particular case but I don't see the general necessity.

The MagicNumber rule already has 11! configuration options and I don't see how your use case justifies yet another one.

**Reply by christian-draeger** [reply]
the when expression is not different in general but i feel like i have read when cases like that hundred of times in a bunch of different code bases, which leads me to believe that this is a common use-case. therefore it would be just nice if it would be possible to not need to clutter the code with @Suppress annotation every time if this particular use-case occurres. nothing more, nothing less :)

i mean a switch statement evaluates an expression, matching the expression's value to a case clause - it is totally valid and natural that these values will be strings, numbers and things like that. this has nothing to do with magic numbers at all imho

**Reply by cortinico** [reply]
I tend to agree with @marschwar here. What makes a `when` statement worth an exception more than a multi-condition if-than-else?

**Reply by radekjezdik** [reply]
Consider this example:

```
when (it) {
    1L -> Type.TARGET
    4L -> Type.IDENTITY
    ...
    else -> error("Unknown type")
}
```

Does it make sense to extract the numbers to constants when it is very clear from the given code? I  think that a lot of mapping codes will be quite similar to this.
